File: rxhttp/src/main/java/rxhttp/wrapper/utils/Converter.java
package rxhttp.wrapper.utils;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import java.io.IOException;
import java.lang.reflect.Type;

import okhttp3.Response;
import okhttp3.ResponseBody;
import rxhttp.Platform;
import rxhttp.wrapper.OkHttpCompat;
import rxhttp.wrapper.callback.IConverter;
import rxhttp.wrapper.entity.ParameterizedTypeImpl;

/**
 * User: ljx
 * Date: 2022/10/30
 * Time: 14:44
 */
public class Converter {

    public static <T> T convertTo(Response response, Type rawType, Type... types) throws IOException {
        return convert(response, ParameterizedTypeImpl.get(rawType, types));
    }

    public static <T> T convertToParameterized(Response response, Type rawType, Type... actualTypes) throws IOException {
        return convert(response, ParameterizedTypeImpl.getParameterized(rawType, actualTypes));
    }

    @SuppressWarnings("unchecked")
    public static <T> T convert(Response response, Type type) throws IOException {
        ResponseBody body = OkHttpCompat.throwIfFail(response);
        if (type == ResponseBody.class) {
            try {
                return (T) OkHttpCompat.buffer(body);
            } finally {
                body.close();
            }
        } else if (Platform.get().isAndroid() && type == Bitmap.class) {
            return (T) BitmapFactory.decodeStream(body.byteStream());
        } else {
            boolean needDecodeResult = OkHttpCompat.needDecodeResult(response);
            IConverter converter = OkHttpCompat.request(response).tag(IConverter.class);
            return converter.convert(body, type, needDecodeResult);
        }
    }
}

File: rxhttp/src/main/java/rxhttp/wrapper/utils/Converter.kt
package rxhttp.wrapper.utils

import okhttp3.Response
import java.io.IOException
import java.lang.reflect.Type
import kotlin.reflect.KClass

/**
 * User: ljx
 * Date: 2020/8/15
 * Time: 11:48
 */
@Throws(IOException::class)
fun <T> Response.convertTo(rawType: KClass<*>, vararg types: Type): T =
    convertTo(rawType.java, *types)

@Throws(IOException::class)
fun <T> Response.convertTo(rawType: Type, vararg types: Type): T =
    Converter.convertTo(this, rawType, *types)

@Throws(IOException::class)
fun <T> Response.convertToParameterized(rawType: KClass<*>, vararg actualTypes: Type): T =
    convertToParameterized(rawType.java, *actualTypes)

@Throws(IOException::class)
fun <T> Response.convertToParameterized(rawType: Type, vararg actualTypes: Type): T =
    Converter.convertToParameterized(this, rawType, *actualTypes)

@Throws(IOException::class)
fun <T> Response.convert(type: Type): T = Converter.convert(this, type)